### DISCUSSION TITLE
LIMIT applied before ORDER BY in complex query with relationships

### DISCUSSION THREAD

**Discussion by scbj** [mikro-orm/mikro-orm#6124] [Q&A]
Hello! 😊

We're encountering an issue when using `findAndCount` with pagination and ordering on a related entity. Specifically, the generated SQL query applies the `LIMIT` before the `ORDER BY` clause within a subquery, leading to incorrect ordering of the results. This happens because the data is paginated before it is sorted.

```typescript
@Entity({ tableName: 'map_metadata' })
@Filter({ name: 'currentVersion', cond: { endDate: null }, default: true })
class MapMetadata {
  @Property()
  title!: string;

  @Property({ columnType: 'timestamptz' })
  startDate;

  @Property({ columnType: 'timestamptz', nullable: true })
  endDate?: Date | null = null;

  @ManyToOne(() => Map)
  map!: Map;
}


@Entity({ tableName: 'map' })
@Filter({ name: 'notDeleted', cond: { deleted: false }, default: true })
class MapV2 extends BaseEntity {
  @Property()
  @Unique()
  @Index({ name: 'accessid_v2_index' })
  accessId!: string;

  @Property({ default: false })
  deleted: boolean = false;

  @OneToMany(() => MapMetadata, (metadata) => metadata.map, {
    orderBy: { startDate: QueryOrder.DESC }
  })
  /** Multiple metadatas because of historization */
  metadata = new Collection<MapMetadata>(this);
}
```

```typescript
this.mapRepository.findAndCount(
  {
    populate: ['metadata'],
    orderBy: {
      metadata: {
        startDate: QueryOrder.DESC,
      },
    },
    limit: 2,
    offset: 0
  },
);
```

To provide some context, whenever a `Map` title is modified, we update the `endDate` of the current `MapMetadata` to the current timestamp, then create a new `MapMetadata` with a `startDate` set to the current timestamp and `endDate` as `null`. This means that for any given `Map`, there could be multiple `MapMetadata` records, with only one having an `endDate` of `null`.

Our goal is to paginate the `Map` entities, while sorting them based on the last modification date (i.e., when the title was last modified). However, MikroORM currently generates SQL queries like this:

```sql
SELECT count(DISTINCT `m0`.`id`) AS `count`
FROM `map` AS `m0`
LEFT JOIN `map_metadata` AS `m1` ON `m0`.`id` = `m1`.`map_id`
WHERE `m0`.`deleted` = false
	AND `m1`.`end_date` IS NULL
```
```sql
SELECT `m0`.*
	,`m1`.`id` AS `m1__id`
	,`m1`.`start_date` AS `m1__start_date`
	,`m1`.`end_date` AS `m1__end_date`
	,`m1`.`title` AS `m1__title`
	,`m1`.`map_id` AS `m1__map_id`
FROM `map` AS `m0`
LEFT JOIN `map_metadata` AS `m1` ON `m0`.`id` = `m1`.`map_id`
	AND `m1`.`end_date` IS NULL
WHERE `m0`.`id` IN (
		SELECT `m0`.`id`
		FROM (
			SELECT `m0`.`id`
			FROM `map` AS `m0`
			LEFT JOIN `map_metadata` AS `m1` ON `m0`.`id` = `m1`.`map_id`
			WHERE `m0`.`deleted` = false
			GROUP BY `m0`.`id`
			ORDER BY Min(`m1`.`start_date`) DESC LIMIT 2
			) AS `m0`
		)
ORDER BY `m1`.`start_date` DESC
``` 

In this case, the `LIMIT` should be applied at the very end of the query, or the condition `m1.end_date IS NULL` should be included in the most nested `WHERE` clause to ensure proper sorting. Without this, the ordering of the results is incorrect. 

We've attempted several approaches, including adding extra filters, changing the `LoadStrategy` to `JOINED`, and other tweaks, but haven't found a solution. We'd prefer not to use the Query Builder to maintain compatibility across multiple databases.

Any insights or suggestions would be greatly appreciated!

**Comment by B4nan** [comment]
Please provide a [complete reproduction](https://github.com/mikro-orm/reproduction), something I can run myself to see the problem. It's hard to debug such issues without the data.

> In this case, the LIMIT should be applied at the very end of the query

You cant apply limit on the root query, since there you get cartesian product for every to-many join you do.
